klippy/extras/bltouch.py

```diff
@@ -53,54 +59,96 @@ def __init__(self, config):
         except self.mcu_endstop.TimeoutError as e:
-            raise homing.EndstopError("BLTouch sensor test failed")
+            raise homing.EndstopError("BLTouch failed to %s" % (msg,))
         for s, pos in zip(self.mcu_endstop.get_steppers(), prev_positions):
             s.set_commanded_position(pos)
+    def raise_probe(self):
+        self.send_cmd('reset')
+        check_start_time = self.send_cmd('pin_up', duration=self.pin_move_time)
+        check_end_time = self.send_cmd(None)
+        self.verify_state(check_start_time, check_end_time,
+                          self.pin_up_triggered, "raise probe")
+    def test_sensor(self):
+        if not self.pin_up_touch_triggered or self.pin_up_triggered:
+            # Nothing to test
+            return
+        toolhead = self.printer.lookup_object('toolhead')
+        print_time = toolhead.get_last_move_time()
+        if print_time < self.next_test_time:
+            self.next_test_time = print_time + TEST_TIME
+            return
+        # Raise the bltouch probe and test if probe is raised
+        self.sync_print_time()
+        check_start_time = self.send_cmd('reset', duration=self.pin_move_time)
+        check_end_time = self.send_cmd('touch_mode')
+        self.send_cmd(None)
+        self.verify_state(check_start_time, check_end_time, True,
+                          "verify sensor state")
         # Test was successful
-        self.next_test_time = home_time + TEST_TIME
-        toolhead.reset_print_time(home_time + 2. * MIN_CMD_TIME)
+        self.next_test_time = check_end_time + TEST_TIME
+        self.sync_print_time()
     def home_prepare(self):
         self.test_sensor()
-        toolhead = self.printer.lookup_object('toolhead')
-        print_time = toolhead.get_last_move_time()
-        self.send_cmd(print_time, 'pin_down')
-        self.send_cmd(print_time + self.pin_move_time, 'touch_mode')
-        toolhead.dwell(self.pin_move_time + MIN_CMD_TIME)
+        self.sync_print_time()
+        self.send_cmd('pin_down', duration=self.pin_move_time - MIN_CMD_TIME)
+        self.send_cmd(None)
+        self.sync_print_time()
         self.mcu_endstop.home_prepare()
+        self.start_mcu_pos = [(s, s.get_mcu_position())
+                              for s in self.mcu_endstop.get_steppers()]
     def home_finalize(self):
-        toolhead = self.printer.lookup_object('toolhead')
-        print_time = toolhead.get_last_move_time()
-        self.send_cmd(print_time, 'reset')
-        self.send_cmd(print_time + MIN_CMD_TIME, 'pin_up')
-        self.send_cmd(print_time + MIN_CMD_TIME + self.pin_move_time, None)
-        toolhead.dwell(self.pin_move_time + MIN_CMD_TIME)
+        self.sync_mcu_print_time()
+        self.raise_probe()
+        self.sync_print_time()
+        # Verify the probe actually deployed during the attempt
+        for s, mcu_pos in self.start_mcu_pos:
+            if s.get_mcu_position() == mcu_pos:
```

Comment:
Does the rest of the code guarantee that we don't move even one step forward in the case that the endstop is triggered before the homing starts?  I'm worried about the following code in `homing.py`
```python
        for mcu_endstop, name in endstops:
            min_step_dist = min([s.get_step_dist()
                                 for s in mcu_endstop.get_steppers()])
            mcu_endstop.home_start(
                print_time, ENDSTOP_SAMPLE_TIME, ENDSTOP_SAMPLE_COUNT,
                min_step_dist / speed)
        # Issue move
        error = None
        try:
            self.toolhead.move(movepos, speed)
```

The toolhead move does not take any printtime, so I assume that it uses the current time. Which means that the endstop detection and move starts at almost the same time. This probably means that the first step is already requested when the endstop trigger is detected, especially with the sample count greater than 1. 

This means that the if `if s.get_mcu_position() == mcu_pos:` fails to detect the error, and the probe homes with the wrong offset.

If you are going to continue using this approach of checking for errors retrospectively, I suggest that you add a parameter for the minimum move distance, by default that should be around 5mm. That parameter would also possibly make it possible to use slightly shorter pin move times, and rely on errors being detected during the initial phase of the movement instead. But only as an optimization by the user when everything is working as it should and you want to optimize the mesh bed leveling times.